lib/concurrency/Prefix_Refinement.thy

```diff
@@ -1103,36 +1805,273 @@ lemma prefix_refinement_select:
+    apply (fastforce simp: prefix_refinement_def)
+   apply (fastforce simp: prefix_refinement_def)
+  apply (clarsimp simp: prefix_refinement_returnOk_imp)
+  done
+
+lemma prefix_refinement_alternate1:
+  "prefix_refinement sr iosr iosr rvr AR R P Q a c \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R P Q (a \<sqinter> b) c"
+  apply (subst prefix_refinement_def, clarsimp)
+  apply (drule(6) pfx_refnD2, clarsimp)
+  apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative1)
+  done
+
+lemma prefix_refinement_alternate2:
+  "prefix_refinement sr iosr iosr rvr AR R P Q b c \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R P Q (a \<sqinter> b) c"
+  apply (subst prefix_refinement_def, clarsimp)
+  apply (drule(6) pfx_refnD2, clarsimp)
+  apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative2)
+  done
+
+lemma prefix_refinement_either_alternate1:
+  "\<lbrakk>prefix_refinement sr iosr iosr rvr AR R P Q a c; prefix_refinement sr iosr iosr rvr AR R P' Q b c\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R (P or P') Q (a \<sqinter> b) c"
+  apply (subst prefix_refinement_def, clarsimp simp del: imp_disjL)
+  apply (erule disjE)
+   apply (drule(6) pfx_refnD2, clarsimp)
+   apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative1)
+  apply (drule(6) pfx_refnD2, clarsimp)
+  apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative2)
+  done
+
+lemma prefix_refinement_either_alternate2:
+  "\<lbrakk>prefix_refinement sr iosr iosr rvr AR R P Q a c; prefix_refinement sr iosr iosr rvr AR R P Q' b c\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R P (Q or Q') (a \<sqinter> b) c"
+  apply (subst prefix_refinement_def, clarsimp simp del: imp_disjL)
+  apply (erule disjE)
+   apply (drule(6) pfx_refnD2, clarsimp)
+   apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative1)
+  apply (drule(6) pfx_refnD2, clarsimp)
+  apply (fastforce intro: triv_refinement_trans[rotated] triv_refinement_alternative2)
+  done
+
+lemma is_matching_fragment_restrict:
+  "is_matching_fragment sr osr rvr ctr cres s0 R s f
+   \<Longrightarrow> is_matching_fragment sr osr rvr ctr cres s0 R s (\<lambda>s'. if s'=s then f s else {})"
+  by (simp add: is_matching_fragment_def prefix_closed_def self_closed_def env_closed_def
+                matching_tr_pfx_def matching_tr_def)
+
+lemma triv_refinement_restrict:
+  "triv_refinement f (\<lambda>s'. if s'=s then f s else {})"
+  by (clarsimp simp: triv_refinement_def)
+
+\<comment> \<open>FIXME: corres rules for condition don't exist, so maybe we don't bother with this. It feels
+          like it shouldn't be this hard to prove, but providing correct fragments is frustrating.
+          It might make it easier to instantiate if the exists was wrapped in a new definition.\<close>
+lemma prefix_refinement_condition_strong:
```

Comment:
`condition` is mostly used in AutoCorres, so there is probably no need to add a complete rule set for it here.

Reply:
I might keep it for now, as an example of a difficult proof which might guide improvements to the definition of `prefix_refinement` and its helper lemmas.